Frameworks/CoreText/DWriteWrapper_CoreText.mm

```diff
@@ -522,4 +523,56 @@ HRESULT STDMETHODCALLTYPE GetPixelsPerDip(_In_opt_ void* clientDrawingContext, _
     }
 
     return frame;
+}
+
+static CGSize _DWriteGetSize(CFAttributedStringRef string, CFRange range, CGSize maxSize, CFRange* fitRange) {
+    CGSize ret = CGSizeZero;
+    if (range.length == 0L) {
+        range.length = CFAttributedStringGetLength(string);
+    }
+
+    if (!string || range.length <= 0L) {
+        return ret;
+    }
+
+    ComPtr<IDWriteTextLayout> textLayout;
+    if (FAILED(__DWriteTextLayoutCreate(string, range, { CGPointZero, maxSize }, &textLayout))) {
+        return ret;
+    }
+
+    DWRITE_TEXT_METRICS textMetrics;
+    if (FAILED(textLayout->GetMetrics(&textMetrics))) {
+        return ret;
+    }
+
+    // TODO:: find more precise value than 1.0 to increase width by to fully enclose frame
+    ret.width = std::min(maxSize.width, textMetrics.widthIncludingTrailingWhitespace + 1.0f);
+    ret.height = std::min(maxSize.height, textMetrics.height);
+
+    if (fitRange) {
+        *fitRange = {};
+        uint32_t lineCount = 0;
+
+        // Ignore return value, we only want the lineCount for now
+        textLayout->GetLineMetrics(nullptr, 0, &lineCount);
+
+        std::vector<DWRITE_LINE_METRICS> metrics(lineCount);
```

Comment:
>::vector<DWRITE_LINE_METRICS> metrics(lineCount); [](start = 11, length = 49)

since this is about performance, use array rather than a vector here (you will have a slight gain)